File: 0084-largest-rectangle-in-histogram/0084-largest-rectangle-in-histogram.java
class Solution {
    public int largestRectangleArea(int[] heights) {
        int area;
		int maxArea = 0;
		int i = 0;
		Stack<Integer> s = new Stack<>();
		while(i < heights.length)
        {
			if(s.isEmpty() || heights[s.peek()] <= heights[i])
            {
				s.push(i);
				i++;
			}
			else
            {
				int temp = s.pop();
				// s.pop();
				if(s.isEmpty())
                {
					area = heights[temp] * i;
				}
				else
                {
					area = heights[temp] * (i - s.peek() - 1);
				}
				if(area > maxArea)
                {
					maxArea = area;
				}
			}
		}
		while(!s.isEmpty())
        {
			int temp = s.pop();
			// s.pop();
			if(s.isEmpty())
            {
				area = heights[temp] * i;
			}
			else
            {
				area = heights[temp] * (i - s.peek() - 1);
			}
			if(area > maxArea)
            {
				maxArea = area;
			}
		}
		return maxArea;
    }
}

File: 0084-largest-rectangle-in-histogram/0084-largest-rectangle-in-histogram.cpp
class Solution {
public:
    int largestRectangleArea(vector<int>& histo) 
    {
        stack <int> st;
        int maxA = 0;
        int n = histo.size();
        for (int i = 0; i <= n; i++) {
            while (!st.empty() && (i == n || histo[st.top()] >= histo[i])) {
            int height = histo[st.top()];
            st.pop();
            int width;
            if (st.empty())
                width = i;
            else
                width = i - st.top() - 1;
            maxA = max(maxA, width * height);
            }
            st.push(i);
        }
        return maxA;
    }
};



// class Solution {
// public:
//     int largestRectangleArea(vector<int>& heights) 
//     {
//         int area;
// 		int maxArea = 0;
// 		int i = 0;
// 		stack<int> s;
// 		while(i < heights.size())
//         {
// 			if(s.empty() or heights[s.top()] <= heights[i])
//             {
// 				s.push(i);
// 				i++;
// 			}
// 			else
//             {
// 				int temp = s.top();
// 				s.pop();
// 				if(s.empty())
//                 {
// 					area = heights[temp] * i;
// 				}
// 				else
//                 {
// 					area = heights[temp] * (i - s.top() - 1);
// 				}
// 				if(area > maxArea)
//                 {
// 					maxArea = area;
// 				}
// 			}
// 		}
// 		while(!s.empty())
//         {
// 			int temp = s.top();
// 			s.pop();
// 			if(s.empty())
//             {
// 				area = heights[temp] * i;
// 			}
// 			else
//             {
// 				area = heights[temp] * (i - s.top() - 1);
// 			}
// 			if(area > maxArea)
//             {
// 				maxArea = area;
// 			}
// 		}
// 		return maxArea;
//     }
// };